docs/content/sips/0xx-spin-factors.md

```diff
@@ -0,0 +1,129 @@
+title = "SIP 0XX - Spin Factors"
+template = "main"
+date = "2024-05-20T12:00:00Z"
+
+---
+
+Summary: Refactor Spin Runtime Functionality with "Factors"
+
+Owner: lann.martin@fermyon.com
+
+Created: May 20, 2024
+
+## Background
+
+Spin 1.0 shipped with a mechanism for extending the runtime environment with
+loosely-coupled host functionality called
+[`HostComponent`](https://fermyon.github.io/rust-docs/spin/v2.2.0/spin_core/trait.HostComponent.html).
+As Spin has evolved, more and more runtime functionality has fallen outside of
+the scope of this mechanism, leading to the code for these features spreading
+out over the Spin codebase. This not only makes it hard to read and understand
+these features but also raises the bar for unfamiliar developers trying to add
+new features.
+
+Separately, the introduction of the SpinKube project has made it clear that
+there will be (at least) two major embeddings of the Spin runtime environment:
+the Spin CLI (`spin up`) and SpinKube (`containerd-shim-spin`). In order to
+provide better integration into the Kubernetes ecosystem, some runtime feature
+implementations will diverge between these embeddings, making the loose coupling
+of features even more important.
+
+## Proposal
+
+The basic inversion of control approach used by `HostComponent`s will be
+redesigned and expanded into a new system called Spin Factors, where independent
+feature sets are organized into individual "factors". A "factor" encapsulates a
+single logical Spin runtime feature into a reusable Rust crate, using the Spin
+Factors framework to "hook" into various points in the lifecycle of a Spin
+trigger's execution.
+
+Some of the goals of this new system:
```

Comment:
Does this get us closer to being able to use RPC host components (rather than embedded Spin ones) too?

Reply:
@kate-goldenring what do you mean when you say RPC host components. I'm just out of the loop and curious 😄.

Reply:
Host components that are not in the spin binary rather run separately and are used over RPC

Reply:
Now I too am curious what the answer to this question is haha

Reply:
For interfaces known at build time an rpc implementation could just be a normal factor. For interfaces known only at runtime it would require hooking into the (pre)instantiation step, which could be added as a factors phase or implemented outside of factors.

Reply:
So I think the answer to the original question is "no but it shouldn't make it harder either".